### DISCUSSION TITLE
Open url in a specific browser profile

### DISCUSSION THREAD

**Discussion by MscrmTools** [MscrmTools/XrmToolBox#1137] [Ideas]
I don't know how you work with multiple customers but I personaly use browser profiles : one profile per customer. So far so good!

Problem with XrmToolBox is that when opening an url it uses the default browser with the default profile, which does not have necessarily the appropriate credentials to open the url.

What I propose is to allow user to select a browser and a profile that can be associated with an XrmToolBox connection. 

When a connection is validated, you will be proposed to set a browser
![image](https://user-images.githubusercontent.com/10774317/160996729-950fad2e-feb0-4a28-b1b0-8191c8f94c15.png)

Then you can select the browser and the profile to use
![image](https://user-images.githubusercontent.com/10774317/160996783-990151a3-5cf2-4094-be1a-609cc9b7eabd.png)

I do have a question for developers that will use this new feature. How should I expose the method to open an Url?

1/ in the `ConnectionDetail` object ? => `ConnectionDetail.OpenUrl(string url)`

2/ in an helper class in `XrmToolBox.Extensibility` namespace ? =>  `XrmToolBox.Extensibility.BrowserHelper.OpenUrl(ConnectionDetail detail)`

**Comment by MarkMpn** [comment]
I'd definitely use this in SQL 4 CDS. Don't really have a strong opinion on which method you use to expose it though.

**Comment by MscrmTools** [comment]
Well, it's implemented the easiest way to avoid any dependency between connection controls and XrmToolBox => I implemented `connectionDetail.OpenUrlWithBrowserProfile(new Uri(url));`

**Reply by rappen** [reply]
If the method was added to XTB instead of ConnectionDetail, we could always call it if we are connected or not.
It may also work if it is added as an extension method for ConnectionDetail, we could always use the method as you example above, even if it is null.
Currently we have to check it we are connected or not, and either call your method 👆 or just call Process.Start(...)

**Reply by MscrmTools** [reply]
As the choice of browser is stored in the connection, then it seems normal to me that you need to be connected to use this feature.
And this is where it's interesting  because you may want to open a page for a specific tenant when connected.
As you said, if you need to open an url with your own parameters, you can always use Process.Start

**Reply by rappen** [reply]
I understand. I just mean that it would be easier if I can open any URL in one and only one way. If I am connected, use the possibly selected browser profile, and if I'm not set on the connection OR not connected at all, used in the same way.
If this "if/or" is handled more close to the core, instead of needing to implement it in each and every tool.
That's my thought :)

**Reply by MscrmTools** [reply]
Thank you for your thoughts!
Not sure it is easily doable. If you want the openUrl to be available anywhere, it means we need a static method. 
If the method is static, I'm not sure we can know what is the current connection. So, we would need to pass the connection where we need to open an url. 
And we are back to a method that needs the connection so my implementation seems logical so far (except if we are not connected)
But prove me I'm wrong and I would be happy to change the way I implemented it.

**Reply by rappen** [reply]
@daryllabar has told me (and everyone else) to use like this:
```
public static void OpenUrlWithBrowserProfile(this ConnectionDetail detail, Uri uri)
```
So it looks like your method, but it works even if my ConnectionDetail is actually null.
Check this first in the method:
```
if (detail == null)
{
  Process.Start(uri.ToString());
}
else
{
  your code
}
```

Something like this...

**Reply by MscrmTools** [reply]
Got it, I can and will add both way to open url

**Reply by daryllabar** [reply]
Wait, what did I do?

**Reply by rappen** [reply]
@daryllabar - you like extensions, just like me 😊

**Reply by MscrmTools** [reply]
Well, just thought about it, extension does not work for null values... does it?

**Reply by rappen** [reply]
I think it works just fine...
It "looks like" a normal method, but it is a static method and you can call it anyway if the object (`ConnectionDetail` in this case) is null of a proper value.
@daryllabar, please explain! :)

**Reply by rappen** [reply]
Check this:
```
        internal static string Value(this TreeNode node, string key)
        {
            if (node != null && node.Tag != null && node.Tag is Dictionary<string, string> tag && tag.ContainsKey(key))
            {
                return tag[key];
            }
            return string.Empty;
        }
```
can be called like this:
```
var alias = node.Value("alias");
```
The "node" can be null of a `TreeNode`, and the method will handle it.
If you add that method to the class, it would explode if you call it when node is null.

**Reply by MscrmTools** [reply]
ok, got it... But you need to have a ConnectionDetail object (wheter it is null or not) available where you want to open an url. 
Whereas a static method that takes a ConnectionDetail as an optional parameter can work anywhere

**Reply by rappen** [reply]
We always have the ConnectionDetail object, it is from the `PluginControlBase`.
I just think it's easier in our code to be possibly to call `ConnectionDetail.OpenUrlWithBrowserProfile(...);` to open an URL. To centralize the checking, both browser/profile and connected or not.

But you're the boss! 😊

**Comment by daryllabar** [comment]
What if I use different browsers as well?  Would that be supported?

**Reply by MscrmTools** [reply]
I handle chrome, edge and Firefox
